server/src/session_manager.ts

```diff
@@ -147,7 +207,135 @@ export class SessionManager {
                 return undefined;
         }
     }
-    // mostly copied from https://github.com/LuanRT/BgUtils/tree/main/examples/node
+
+    private getFetch(dispatcher: Agent | undefined): FetchFunction {
+        return async (url: any, options: any): Promise<any> => {
+            const maxRetries = 3;
+            for (let attempts = 1; attempts <= maxRetries; attempts++) {
+                try {
+                    const response = await axios.post(url, options.body, {
+                        headers: options.headers,
+                        httpsAgent: dispatcher,
+                    });
+
+                    return {
+                        ok: true,
+                        json: async () => {
+                            return response.data;
+                        },
+                    };
+                } catch (e) {
+                    if (attempts >= maxRetries) {
+                        return {
+                            ok: false,
+                            json: async () => {
+                                return null;
+                            },
+                            status: e.response?.status || e.code,
+                        };
+                    }
+                    await new Promise((resolve) => setTimeout(resolve, 5000));
+                }
+            }
+        };
+    }
+
+    private async generateBotGuardData(
+        pxySpec: ProxySpec,
+        bgClient: BG.BotGuardClient,
+    ): Promise<BGData> {
+        try {
+            const doFetch = this.getFetch(this.getProxyDispatcher(pxySpec));
+            const webPoSignalOutput: WebPoSignalOutput = [];
+            const botguardResponse = await bgClient.snapshot({
+                webPoSignalOutput,
+            });
+            const integrityTokenResp = await doFetch(buildURL("GenerateIT"), {
+                method: "POST",
+                headers: getHeaders(),
+                body: JSON.stringify([
+                    SessionManager.REQUEST_KEY,
+                    botguardResponse,
+                ]),
+            });
+
+            const [
+                integrityToken,
+                estimatedTtlSecs,
+                mintRefreshThreshold,
+                websafeFallbackToken,
+            ] = (await integrityTokenResp.json()) as [
+                string,
+                number,
+                number,
+                string,
+            ];
+
+            const integrityTokenData = {
+                integrityToken,
+                estimatedTtlSecs,
+                mintRefreshThreshold,
+                websafeFallbackToken,
+            };
+
+            if (!integrityToken)
+                throw new Error(
+                    `Unexpected empty integrity token, response: ${JSON.stringify(integrityTokenData)}`,
+                );
+            const bgData: BGData = {
+                cachedTokenMinter: {
+                    expiry: new Date(Date.now() + estimatedTtlSecs * 1000),
```

Comment:
With a function like this:

```ts
function addToNow(
    hours: number = 0,
    minutes: number = 0,
    seconds: number = 0,
    startDate?: Date
): Date {
    const future = startDate ? new Date(startDate) : Date.now();
    future.setHours(hours + future.getHours());
    future.setMinutes(minutes + future.getMinutes());
    future.setSeconds(seconds + future.getSeconds());
    return future;
}
```


```suggestion
                    expiry: addToNow(0, 0, estimatedTtlSecs),
```

Reply:
Again, I'd rather keep our current one because this line will soon be removed